services/vision/obstacledepth/obstacle_depth.go

```diff
@@ -0,0 +1,376 @@
+		return nil, errors.New("config for obstacle_depth cannot be nil")
+	}
+
+	// If you have no intrinsics,
+	if conf.intrinsics == nil {
+		r.Logger().Warn("obstacle depth started without camera's intrinsic parameters")
+		segmenter := buildObsDepthNoIntrinsics()
+		return svision.NewService(name, r, nil, nil, nil, segmenter)
+	}
+
+	// Use defaults if needed
+	if conf.Hmax == 0 {
+		conf.Hmax = defaultHmax
+	}
+	if conf.ThetaMax == 0 {
+		conf.ThetaMax = defaultThetamax
+	}
+
+	myObsDep := obsDepth{
+		hMin: conf.Hmin, hMax: conf.Hmax, sinTheta: math.Sin(conf.ThetaMax),
+		intrinsics: conf.intrinsics, returnPCDs: conf.ReturnPCDs, k: conf.K,
+	}
+	segmenter := myObsDep.buildObsDepthWithIntrinsics() // does the thing
+	return svision.NewService(name, r, nil, nil, nil, segmenter)
+}
+
+// buildObsDepthNoIntrinsics will return the shortest depth in the depth map as a Geometry point.
+func buildObsDepthNoIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		min, _ := dm.MinMax()
+
+		pt := spatialmath.NewPoint(r3.Vector{X: 0, Y: 0, Z: float64(min)}, "obstacle")
+		toReturn := make([]*vision.Object, 1)
+		toReturn[0] = &vision.Object{Geometry: pt}
+
+		return toReturn, nil
+	}
+}
+
+// buildObsDepthWithIntrinsics will use the methodology in Manduchi et al. to find obstacle points
+// before clustering and projecting those points into 3D obstacles.
+func (o *obsDepth) buildObsDepthWithIntrinsics() segmentation.Segmenter {
+	return func(ctx context.Context, src camera.VideoSource) ([]*vision.Object, error) {
+		// Check if we have intrinsics here or in the camera properties. If not, don't even try
+		props, err := src.Properties(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if props.IntrinsicParams == nil && o.intrinsics == nil {
+			return nil, errors.New("tried to build obstacles depth with intrinsics but no instrinsics found")
+		} else if props.IntrinsicParams != nil {
+			o.intrinsics = props.IntrinsicParams
+		}
+
+		depthStream, err := src.Stream(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get stream from %s", src)
+		}
+		pic, release, err := depthStream.Next(ctx)
+		if err != nil {
+			return nil, errors.Errorf("could not get image from stream %s", depthStream)
+		}
+		defer release()
+		dm, err := rimage.ConvertImageToDepthMap(ctx, pic)
+		if err != nil {
+			return nil, errors.New("could not convert image to depth map")
+		}
+		o.dm = dm
+		o.makePointList(sampleN)
+		runtime.GOMAXPROCS(maxThreads)
```

Comment:
This will set the max CPUs for all of viam-server, which we should not do in this context. Also, I don't believe this will limit the number of goroutines, since there can be many more goroutines than CPUs.

Reply:
Okok gone

Reply:
Thanks. I like the idea of making the parallelism configurable somehow though. Maybe chunkSize and sampleN should be in the config?